### PR TITLE
UefiCpuPkg/PiSmmCpuDxeSmm:Check resource HOB range before mapping

### PR DESCRIPTION
**Pull request by td36**
# Description

This commit is to check if the resource HOB range does not exceed the max supported physical address.
The function BuildMemoryMapFromResDescHobs is to build Memory Region from resource HOBs. Then the memory maps will be used during creating or modifying SMM page table. If the resource HOB range exceeds the max supported physical address, then subsequent calling of PageTableMap() will fail.

## How This Was Tested
Tested in Intel internal server platform with x86 standalone MM enabled and succussfully booted into shell

### CODE REVIEW

**Review by kraxel — APPROVED** [review]

File: UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c [hunk 1]
```diff
@@ -138,11 +141,22 @@ BuildMemoryMapFromResDescHobs (
...
+        // Logging attribute range is treated as not present. Not-present ranges are not included in this memory map.
+        //
+        Count++;
+      } else {
+        DEBUG ((
+          DEBUG_ERROR,
+          "The resource HOB range [0x%lx, 0x%lx] which exceeds the max supported physical address won't be mapped.",
+          Hob.ResourceDescriptor->PhysicalStart,
+          ResourceHobEnd
+          ));
```
Comment (jiaxinwu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c:158]: Add assert here

File: UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c [hunk 2]
```diff
@@ -138,11 +141,22 @@ BuildMemoryMapFromResDescHobs (
...
   while (Hob.Raw != NULL) {
     if ((Hob.ResourceDescriptor->ResourceAttribute & MM_RESOURCE_ATTRIBUTE_LOGGING) == 0) {
-      //
-      // Resource HOBs describe all accessible non-smram regions.
-      // Logging attribute range is treated as not present. Not-present ranges are not included in this memory map.
-      //
-      Count++;
+      ResourceHobEnd = Hob.ResourceDescriptor->PhysicalStart + Hob.ResourceDescriptor->ResourceLength;
+
+      if (ResourceHobEnd <= MaxPhysicalAddress) {
```
Comment (jiaxinwu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c:146]: Just need assert:
ASSERT (Hob.ResourceDescriptor->PhysicalStart < MaxPhysicalAddress && ResourceHobEnd < MaxPhysicalAddress)

File: UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c [hunk 3]
```diff
@@ -158,15 +172,19 @@ BuildMemoryMapFromResDescHobs (
   Hob.Raw = GetFirstHob (EFI_HOB_TYPE_RESOURCE_DESCRIPTOR);
   while (Hob.Raw != NULL) {
     if ((Hob.ResourceDescriptor->ResourceAttribute & MM_RESOURCE_ATTRIBUTE_LOGGING) == 0) {
-      ASSERT (Index < Count);
-      (*MemoryRegion)[Index].Base      = Hob.ResourceDescriptor->PhysicalStart;
-      (*MemoryRegion)[Index].Length    = Hob.ResourceDescriptor->ResourceLength;
-      (*MemoryRegion)[Index].Attribute = EFI_MEMORY_XP;
-      if (Hob.ResourceDescriptor->ResourceAttribute == EFI_RESOURCE_ATTRIBUTE_READ_ONLY_PROTECTED) {
-        (*MemoryRegion)[Index].Attribute |= EFI_MEMORY_RO;
+      ResourceHobEnd = Hob.ResourceDescriptor->PhysicalStart + Hob.ResourceDescriptor->ResourceLength;
```
Comment (jiaxinwu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c:175]: don't need change here.

File: UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c (hunk 2 above)
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c:146]: what if the unexpected behavior happens in release build?

File: UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c (hunk 3 above)
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c:175]: You mean remove line 175 directly use the variable in line 177? I'm fine with it

File: UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c (hunk 2 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c:146]: for any issue check, we will use the debug version. Here, it's just help us quick find the problem. debug log will also not output in release version.

File: UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c (hunk 3 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c:175]: i mean once we assert, those code are not needed.

**Review by jiaxinwu — APPROVED** [review]

**Review by niruiyu — APPROVED** [review]

**Pull request merged**